subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingConfigurationBuildOptionFactory.java

```diff
@@ -51,104 +52,130 @@ public LoggingConfigurationBuildOptionFactory() {
+            try {
+                LogLevel logLevel = LogLevel.valueOf(value.toUpperCase());
+                if (logLevel == LogLevel.ERROR) {
+                    throw new IllegalArgumentException("Log level cannot be set to 'ERROR'.");
+                }
+                return logLevel;
+            } catch (IllegalArgumentException e) {
+                String message = String.format("Value '%s' given for %s system property is invalid.  (must be one of quiet, warn, lifecycle, info, or debug)", value, gradleProperty);
+                throw new IllegalArgumentException(message, e);
+            }
         }
     }
 
-    public static class StacktraceOption extends NoArgumentBuildOption<LoggingConfiguration> {
-        public static final String LONG_OPTION = "stacktrace";
-        public static final String SHORT_OPTION = "s";
+    public static class StacktraceOption extends AbstractBuildOption<LoggingConfiguration> {
+        public static final String STACKTRACE_LONG_OPTION = "stacktrace";
+        public static final String STACKTRACE_SHORT_OPTION = "s";
+        public static final String FULL_STACKTRACE_LONG_OPTION = "full-stacktrace";
+        public static final String FULL_STACKTRACE_SHORT_OPTION = "S";
 
         public StacktraceOption() {
-            super(null, CommandLineOptionConfiguration.create(LONG_OPTION, SHORT_OPTION, "Print out the stacktrace for all exceptions."));
+            super(null, CommandLineOptionConfiguration.create(STACKTRACE_LONG_OPTION, STACKTRACE_SHORT_OPTION, "Print out the stacktrace for all exceptions."), CommandLineOptionConfiguration.create(FULL_STACKTRACE_LONG_OPTION, FULL_STACKTRACE_SHORT_OPTION, "Print out the full (very verbose) stacktrace for all exceptions."));
         }
 
         @Override
-        public void applyTo(LoggingConfiguration settings) {
-            settings.setShowStacktrace(ShowStacktrace.ALWAYS);
+        public void applyFromProperty(Map<String, String> properties, LoggingConfiguration settings) {
+            // not supported
         }
-    }
 
-    public static class FullStacktraceOption extends NoArgumentBuildOption<LoggingConfiguration> {
-        public static final String LONG_OPTION = "full-stacktrace";
-        public static final String SHORT_OPTION = "S";
+        @Override
+        public void configure(CommandLineParser parser) {
+            for (CommandLineOptionConfiguration config : commandLineOptionConfigurations) {
+                configureCommandLineOption(parser, config.getAllOptions(), config.getDescription(), config.getDeprecationWarning(), config.isIncubating());
+            }
 
-        public FullStacktraceOption() {
-            super(null, CommandLineOptionConfiguration.create(LONG_OPTION, SHORT_OPTION, "Print out the full (very verbose) stacktrace for all exceptions."));
+            List<String> allShortOptions = CollectionUtils.collect(commandLineOptionConfigurations, new Transformer<String, CommandLineOptionConfiguration>() {
```

Comment:
The imperative equivalent would be shorter and more readable (i.e. add the short options to a list in the for-loop above.

Reply:
done